src/bufferer.rs

```diff
@@ -1,294 +1,587 @@
-            .to_string();
+    fn read_utf16_string(&mut self, length: usize) -> GDResult<String> {
+        let slice = self.read_slice(length * 2)?;
+        let u16_data = slice
+            .chunks_exact(2)
+            .map(|chunk| T::read_u16(chunk))
+            .collect::<Vec<u16>>();
 
-        self.move_position_ahead(value.len() + 1);
-        Ok(value)
+        // ! map error to GDResult
+        String::from_utf16(&u16_data).map_err(todo!())
     }
+}
 
-    pub fn get_string_utf8_unended(&mut self) -> GDResult<String> {
-        let sub_buf = self.remaining_data();
-        if sub_buf.is_empty() {
-            return Err(PacketUnderflow);
-        }
+// Macro Usage
+//
+// let mut deserializer =
+// ByteArrayDeserializer::<LittleEndian>::new(&data);
+//
+// let value_u8: u8 = read_number!(deserializer, u8)?;
+// let value_u16: u16 = read_number!(deserializer, u16)?;
+// let value_u32: u32 = read_number!(deserializer, u32)?;
+// let value_u64: u64 = read_number!(deserializer, u64)?;
+// let value_i8: i8 = read_number!(deserializer, i8)?;
+// let value_i16: i16 = read_number!(deserializer, i16)?;
+// let value_i32: i32 = read_number!(deserializer, i32)?;
+// let value_i64: i64 = read_number!(deserializer, i64)?;
+// let value_f32: f32 = read_number!(deserializer, f32)?;
+// let value_f64: f64 = read_number!(deserializer, f64)?;
+#[allow(unused_macros)]
+macro_rules! read_number {
+    ($deserializer:expr, u8) => {
+        $deserializer.read_u8()
+    };
+    ($deserializer:expr, u16) => {
+        $deserializer.read_u16()
+    };
+    ($deserializer:expr, u32) => {
+        $deserializer.read_u32()
+    };
+    ($deserializer:expr, u64) => {
+        $deserializer.read_u64()
+    };
+    ($deserializer:expr, i8) => {
+        $deserializer.read_i8()
+    };
+    ($deserializer:expr, i16) => {
+        $deserializer.read_i16()
+    };
+    ($deserializer:expr, i32) => {
+        $deserializer.read_i32()
+    };
+    ($deserializer:expr, i64) => {
+        $deserializer.read_i64()
+    };
+    ($deserializer:expr, f32) => {
+        $deserializer.read_f32()
+    };
+    ($deserializer:expr, f64) => {
+        $deserializer.read_f64()
+    };
+}
 
-        let value = std::str::from_utf8(sub_buf)
-            .map_err(|_| PacketBad)?
-            .to_string();
+// Macro Usage
+//
+// let utf8_string: String = read_string!(deserializer, utf8, 10)?;
+// let utf16_string: String = read_string!(deserializer, utf16, 5)?;
+//
+// --------------------
+//
+// #[allow(unused_macros)]
+// macro_rules! read_string {
+// ($deserializer:expr, utf8, $length:expr) => {
+// $deserializer.read_utf8_string($length)
+// };
+// ($deserializer:expr, optional_utf8, $length:expr) => {
+// $deserializer.read_optional_utf8_string($length)
+// };
+// ($deserializer:expr, remaining_utf8) => {
+// $deserializer.read_remaining_utf8_string()
+// };
+// ($deserializer:expr, utf16, $length:expr) => {
+// $deserializer.read_utf16_string($length)
+// };
+// }
```

Comment:
The problem with getting strings (at least with utf8) is that we might want to:
1. get the remaining data as a string
2. get a string from the data (the end is a null byte)
3. look for a string, if there is one, return it, if not, dont (no null byte)

and i guess we can say that this was the initial incentive into rethinking the 'bufferer'.
This is just a note for your further strings implementation.